### PR TITLE
Adding Archer Integration

### PR DESCRIPTION
**Pull request by maiorrulz**

### CODE REVIEW

**Comment by maiorrulz** [comment]
For now w/o fetch

File: Integrations/integration-ArcherRSA.yml [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+            if(index !== -1) {
+                searchRes = searchRes.substr(index + 1);
+            }
+        }
+        var parsedJson = JSON.parse(x2j(searchRes));
+        var records = parsedJson.Records.Record;
+        if(!records) {
+            return [];
+        }
+        console.log("before forEach : " + JSON.stringify(records))
```
Comment (meirwah) [inline Integrations/integration-ArcherRSA.yml]: why all this console.log?

File: Integrations/integration-ArcherRSA.yml [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+    if (result.IsSuccessful != true) {
+        throw 'Login failed. Result = ' + JSON.stringify(result);
+    }
+
+    var token = result.RequestedObject.SessionToken;
+    console.log("Token =" + token);
+    var commandArgs = {};
+    switch (command) {
+        case 'test-module':
+            result = sendRequest("logout", urlDictionary.logout, token, {Value: token});
```
Comment (meirwah) [inline Integrations/integration-ArcherRSA.yml]: test does logout req???

File: Integrations/integration-ArcherRSA.yml [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+        },
+        false);
+
+    if (result.IsSuccessful != true) {
+        throw 'Login failed. Result = ' + JSON.stringify(result);
+    }
+
+    var token = result.RequestedObject.SessionToken;
+    console.log("Token =" + token);
+    var commandArgs = {};
```
Comment (meirwah) [inline Integrations/integration-ArcherRSA.yml]: why is this global ? is no reason for that...

File: Integrations/integration-ArcherRSA.yml [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+            return getListOfApplications(command, token);
+        case 'archer-create-content':
+            return createContent(command, token , args);
+        case 'archer-update-content':
+            return updateContent(command, token , args);
+        case 'archer-get-contents-by-appId':
+            return getContentsByAppId(command, token , args);
+        case 'fetch-incidents':
+            commandArgs = createContantSoapRequest(token, 433, 'demo');
+            return fetchIncidentFlow();
```
Comment (meirwah) [inline Integrations/integration-ArcherRSA.yml]: can't you just pass commandArgs as argument?

File: Integrations/integration-ArcherRSA.yml [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+        case 'archer-list-applications':
+            return getListOfApplications(command, token);
+        case 'archer-create-content':
+            return createContent(command, token , args);
+        case 'archer-update-content':
+            return updateContent(command, token , args);
+        case 'archer-get-contents-by-appId':
+            return getContentsByAppId(command, token , args);
+        case 'fetch-incidents':
+            commandArgs = createContantSoapRequest(token, 433, 'demo');
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: I guess we don't want 433 hardcoded

File: Integrations/integration-ArcherRSA.yml [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+             var fields = record.Field;
+             if(!Array.isArray(fields)) {
+                 fields = [fields];
+             }
+             fields.forEach(function (field) {
+                 el.Fields[mapping[field['-id']].Name] = field['#text'];
+             });
+            outputs.push(el);
+        });
+        console.log("In parse XML outputs: " + JSON.stringify(outputs));
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: I think you can remove this log

File: Integrations/integration-ArcherRSA.yml [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+        return outputs;
+    }
+
+
+    function sendSoapRequest(cmd, bodyContent, SOAPAction) {
+        var headers;
+        headers = {'Content-Type': ['text/xml; charset=utf-8']};
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: this can be in single line

File: Integrations/integration-ArcherRSA.yml [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+        if(SOAPAction) {
+            headers['SOAPAction'] = [SOAPAction];
+        }
+        var url =  server +  urlDictionary[cmd];
+        console.log('Sending request: '  + url + ' with header: ' + JSON.stringify(headers) + ' and body: ' + JSON.stringify(bodyContent));
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: remove console log

File: Integrations/integration-ArcherRSA.yml [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+            if(parsedId >  now.currentId) {
+                console.log("maxId = " + maxId);
+                incidentRes.push(el);
+                if(maxId < parsedId){
+                    maxId = parsedId;
+                }
+            }
+        });
+        setLastRun({currentId: maxId});
+        return (JSON.stringify(incidentRes));
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: why do you wrap the result with ()?

File: Integrations/integration-ArcherRSA.yml [hunk 10]
```diff
@@ -0,0 +1,626 @@
...
+        '               </Criteria>'+
+        '            </SearchReport> ]]>'+
+        '         </searchOptions>'+
+        '         <pageNumber>1</pageNumber>'+
+        '      </ExecuteSearch>'+
+        '   </soap:Body>'+
+        '</soap:Envelope>';
+    }
+
+    function createContantSoapRequest(token, appId, fields) {
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: Should be createContentSoapRequest

File: Integrations/integration-ArcherRSA.yml [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+        ' </UpdateRecord>' +
+        ' </soap:Body>' +
+        ' </soap:Envelope>'
+    }
+
+    function getLevelsByApplicationId(token) {
+
+        var lvlResponse = sendRestRequest('levelIdByModuleId', '', token, true);
+        var levels = dq(lvlResponse, 'RequestedObject.Id');
+        console.log('---Levels: ' + levels);
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: remove all console logs...

File: Integrations/integration-ArcherRSA.yml [hunk 12]
```diff
@@ -0,0 +1,626 @@
...
+        return levels;
+    }
+
+    function createMappingFieldIdToName(levels, token) {
+        console.log("in createMappingFieldIdToName");
+        var mapping = {};
+        if(!levels) {
+            throw 'Mapping levels to Fields failed, levels = ' + JSON.stringify(levels);
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: level does not exits (can be null) and you are doing JSON.stringify on it.. should be instead:
`throw 'Mapping levels to Fields failed, levels not provided`

File: Integrations/integration-ArcherRSA.yml [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+        });
+        console.log("Mapping:\n" + JSON.stringify(mapping));
+        return mapping;
+    }
+
+    function createMappingNameToFieldId(levels, token) {
+        console.log("in createMappingFieldIdToName");
+        var mapping = {};
+        if(!levels) {
+            throw 'Mapping levels to Fields failed, levels = ' + JSON.stringify(levels);
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: level does not exits (can be null) and you are doing JSON.stringify on it.. should be instead:
`throw 'Mapping levels to Fields failed, levels not provided`

**Comment by dorsha** [comment]
I got the below error when running:
`!archer-get-contents-by-appId applicationId="433" fieldsToDisplay="[\"Title\"]"`

image

**Comment by dorsha** [comment]
Command missing: `archer-list-content` (Get a list of records in an application)

**Comment by dorsha** [comment]
Command missing: `archer-search-records` (Get a list of records in an application).
This command should search for records according to specific search.
It is really important to have, otherwise the flow is broken.. the user need to search for the record to update it.

File: Integrations/integration-ArcherRSA.yml [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+    function getContentsByAppIdRequest (token, appId, displayFields) {
+        return  '<?xml version="1.0" encoding="UTF-8"?>'+
+        '    <soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">'+
+        '       <soap:Body>'+
+        '          <ExecuteSearch xmlns="http://archer-tech.com/webservices/">'+
+        '             <sessionToken>' + token + '</sessionToken>'+
+        '         <searchOptions>'+
+        '            <![CDATA[<SearchReport>'+
+        '               <PageSize>100</PageSize>'+
+        '               <MaxRecordCount>100/MaxRecordCount>'+
```
Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: should be configureable (command level)

Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: and also the closing tag has syntax error.. should be:
`<MaxRecordCount>100</MaxRecordCount>`

Comment (dorsha) [inline Integrations/integration-ArcherRSA.yml]: PageSize should use the same argument as the MaxRecordCount

**Comment by dorsha** [comment]
@maiorrulz I pushed a fix for all the code review stuff beside the fetch and the search.
I also changed the command names and descriptions a bit.

**Comment by dorsha** [comment]
Search and sort implemented.
Only fetch left.

**Comment by dorsha** [comment]
We gonna merge this after having test playbook

**Pull request merged**